Sources/_RegexParser/Regex/Parse/Sema.swift

```diff
@@ -0,0 +1,396 @@
+      // TODO: We should error on unknown Unicode scalar names.
+      break
+
+    case .char, .scalar, .startOfLine, .endOfLine, .any:
+      break
+    }
+  }
+
+  func validateCustomCharacterClass(_ c: AST.CustomCharacterClass) throws {
+    for member in c.members {
+      try validateCharacterClassMember(member)
+    }
+  }
+
+  func validateCharacterClassRange(
+    _ range: AST.CustomCharacterClass.Range
+  ) throws {
+    let lhs = range.lhs
+    let rhs = range.rhs
+
+    try validateAtom(lhs)
+    try validateAtom(rhs)
+
+    guard lhs.isValidCharacterClassRangeBound else {
+      throw error(.invalidCharacterClassRangeOperand, at: lhs.location)
+    }
+    guard rhs.isValidCharacterClassRangeBound else {
+      throw error(.invalidCharacterClassRangeOperand, at: rhs.location)
+    }
+
+    guard lhs.literalCharacterValue != nil else {
+      throw error(
+        .unsupported("character class range operand"), at: lhs.location)
+    }
+
+    guard rhs.literalCharacterValue != nil else {
+      throw error(
+        .unsupported("character class range operand"), at: rhs.location)
+    }
+
+    // TODO: Validate lhs <= rhs? That may require knowledge of case
+    // insensitivity though.
+  }
+
+  func validateCharacterClassMember(
+    _ member: AST.CustomCharacterClass.Member
+  ) throws {
+    switch member {
+    case .custom(let c):
+      try validateCustomCharacterClass(c)
+
+    case .range(let r):
+      try validateCharacterClassRange(r)
+
+    case .atom(let a):
+      try validateAtom(a)
+
+    case .setOperation(let lhs, _, let rhs):
+      for lh in lhs { try validateCharacterClassMember(lh) }
+      for rh in rhs { try validateCharacterClassMember(rh) }
+
+    case .quote, .trivia:
+      break
+    }
+  }
+
+  func validateGroup(_ group: AST.Group) throws {
+    let kind = group.kind
+    switch kind.value {
+    case .capture, .namedCapture, .nonCapture, .lookahead, .negativeLookahead:
+      break
+
+    case .balancedCapture:
+      // These are .NET specific, and kinda niche.
+      throw error(.unsupported("balanced capture"), at: kind.location)
+
+    case .nonCaptureReset:
+      // We need to figure out how these interact with typed captures.
+      throw error(.unsupported("branch reset group"), at: kind.location)
+
+    case .atomicNonCapturing:
+      throw error(.unsupported("atomic group"), at: kind.location)
+
+    case .nonAtomicLookahead:
+      throw error(.unsupported("non-atomic lookahead"), at: kind.location)
+
+    case .lookbehind, .negativeLookbehind, .nonAtomicLookbehind:
+      throw error(.unsupported("lookbehind"), at: kind.location)
+
+    case .scriptRun, .atomicScriptRun:
+      throw error(.unsupported("script run"), at: kind.location)
+
+    case .changeMatchingOptions(let opts):
+      try validateMatchingOptions(opts)
+    }
+    try validateNode(group.child)
+  }
+
+  func validateQuantification(_ quant: AST.Quantification) throws {
+    try validateNode(quant.child)
```

Comment:
Is it possible to validate that the child isn't a zero-width assertion here? e.g. we want to reject `\b+`. Tracked in #312.

Reply:
Implemented logic to check for escape sequences that aren't quantifiable, how does it look?